protocol/communities/manager.go

```diff
@@ -1501,25 +1487,90 @@ func (m *Manager) HandleCommunityRequestToJoin(signer *ecdsa.PublicKey, request
 	return requestToJoin, nil
 }
 
-func (m *Manager) checkPermissionToJoin(permissions []*protobuf.CommunityTokenPermission, walletAddresses []gethcommon.Address) (bool, error) {
+func getRevealedChainIDs(revealedAccounts []*protobuf.RevealedAccount, requirementsChainIDs map[uint64]bool) []uint64 {
 
-	erc20TokenRequirements, erc721TokenRequirements := extractTokenRequirements(permissions)
+	revealedAccountsChainIDs := make([]uint64, 0)
+	revealedAccountsChainIDsMap := make(map[uint64]bool)
 
-	// find owned ERC721 tokens required by community's permissions
-	ownedERC721Tokens, err := m.getOwnedERC721Tokens(walletAddresses, erc721TokenRequirements)
-	if err != nil {
-		return false, err
+	// we want all chainIDs provided by revealed addresses that also exist
+	// in the token requirements
+	for _, revealedAccount := range revealedAccounts {
+		for _, chainID := range revealedAccount.ChainIds {
+			if requirementsChainIDs[chainID] && !revealedAccountsChainIDsMap[chainID] {
+				revealedAccountsChainIDsMap[chainID] = true
+				revealedAccountsChainIDs = append(revealedAccountsChainIDs, chainID)
+			}
+		}
 	}
+	return revealedAccountsChainIDs
+}
 
-	// find owned ERC20 token balances required by community's permissions
-	ownedERC20Tokens, err := m.getAccumulatedTokenBalances(walletAddresses, erc20TokenRequirements)
-	if err != nil {
-		return false, err
+func (m *Manager) checkPermissionToJoin(permissions []*protobuf.CommunityTokenPermission, revealedAccounts []*protobuf.RevealedAccount) (bool, error) {
+
+	erc20TokenRequirements, erc721TokenRequirements, ensRequirements := ExtractTokenCriteria(permissions)
+
+	erc20ChainIDsMap := make(map[uint64]bool)
+	erc721ChainIDsMap := make(map[uint64]bool)
+
+	erc20TokenAddresses := make([]gethcommon.Address, 0)
+	revealedAddresses := make([]gethcommon.Address, 0)
+
+	for _, revealedAccount := range revealedAccounts {
+		revealedAddresses = append(revealedAddresses, gethcommon.HexToAddress(revealedAccount.Address))
 	}
 
-	ownedENSNames, err := m.GetOwnedENS(walletAddresses)
-	if err != nil {
-		return false, err
+	for chainID, tokens := range erc20TokenRequirements {
+		erc20ChainIDsMap[chainID] = true
+		for contractAddress := range tokens {
+			erc20TokenAddresses = append(erc20TokenAddresses, gethcommon.HexToAddress(contractAddress))
+		}
+	}
+
+	for chainID, _ := range erc721TokenRequirements {
+		erc721ChainIDsMap[chainID] = true
+	}
+
+	revealedChainIDsForERC20 := getRevealedChainIDs(revealedAccounts, erc20ChainIDsMap)
+	revealedChainIDsForERC721 := getRevealedChainIDs(revealedAccounts, erc721ChainIDsMap)
+
+	// if there are no chain IDs attached to any of the revealed accounts
+	// we aren't able to check balances on selected networks
+	if len(erc20ChainIDsMap) > 0 && len(revealedChainIDsForERC20) == 0 {
+		return false, nil
+	}
```

Comment:
Should we do the same check for erc721?

Reply:
Actually we might even **remove** this check here.

Here's why:

The chain IDs we're checking on here are based on **all** token criteria from **all** permissions.
This means, it's perfectly fine to have criteria that need erc20 token balances, but not have any erc20 balances in an account **if** there's other criteria that, say, require ERC721 or ENS tokens which do exist in any of the user's accounts.

Remember that only **one** permission's criteria has to be met to be granted access.

So if you don't have any ERC20 tokens, even though there's a criteria in any of the permissions, no big deal, cause you might fulfill the criteria of a different permission.

Hence, we actually shouldn't return here.